aplib.net-demo/Assets/Scripts/Player/Movement.cs

```diff
@@ -0,0 +1,190 @@
+
+    [SerializeField] private float _jumpHeight = 7;
+    [SerializeField] private float _airMovementScale = 0.5f;
+    private bool _jumpPressed;
+
+    [SerializeField] private LayerMask _groundMask;
+    [SerializeField] private float _groundDrag = 0.7f;
+    [SerializeField] private float _slopeAngle = 40;
+    [SerializeField] private float _slopeCheckRayExtraLength = 0.3f;
+    [SerializeField] private float _wallCheckMaxDistance = 0.5f;
+    private float _playerHeight;
+    private float _playerRadius;
+    private Vector3 _bottomPoint;
+    private bool _isGrounded;
+
+    private Rigidbody _rigidbody;
+    private CapsuleCollider _controller;
+
+    /// <summary>
+    /// Initialize the player's rigidbody and collider, and freeze the player's rotation.
+    /// </summary>
+    private void Start() { 
+        _rigidbody = GetComponent<Rigidbody>();
+        _controller = GetComponent<CapsuleCollider>();
+        _playerHeight = _controller.height;
+        _playerRadius = _controller.radius;
+        _rigidbody.freezeRotation = true;
+        _gravity = Physics.gravity.y;
+    }
+
+    /// <summary>
+    /// Draw a red wire sphere at the bottom of the player to visualize the ground check area.
+    /// </summary>
+    void OnDrawGizmosSelected() {
+        Gizmos.color = Color.red;
+        Gizmos.DrawWireSphere(_bottomPoint, _playerRadius);
+    }
+
+    /// <summary>
+    /// Main update loop where we do a ground check, move the player, and handle jumping.
+    /// Updates once per physics update.
+    /// </summary>
+    private void FixedUpdate() {
+        // Ground check by checking a sphere at the bottom of the player, more consistent than ray
+        _bottomPoint = transform.TransformPoint(_controller.center - Vector3.up * (_playerHeight * 0.55f - _playerRadius));
+        _isGrounded = Physics.CheckSphere(_bottomPoint, _playerRadius, _groundMask);
+
+        MovePlayer();
+        HandleJump();
+    }   
+    
+    /// <summary>
+    /// Move the player based on the input and the ground check using AddForce.
+    /// Also applies custom gravity and limits the player's velocity.
+    /// </summary>
+    private void MovePlayer() {
+        _horizontalVelocity = transform.right * _horizontalInput.x + transform.forward * _horizontalInput.y;
+        bool isOnSlope = OnWalkableSlope(out Vector3 directionOnSlope, out RaycastHit downHit);
+        bool isWallColliding = WalkingAgainstWall(out Vector3 wallNormal);
+        _rigidbody.useGravity = !isOnSlope;
+
+        if (_isGrounded) {
+            _rigidbody.drag = _groundDrag; 
+            
+            if (isOnSlope) {
+                _rigidbody.AddForce(_maxSpeed * _acceleration * Time.fixedDeltaTime * directionOnSlope);
+                
+                // Apply gravity but perpendicular to the slope, to prevent sliding
+                _rigidbody.velocity += (_actualGravityScale - 1) * _gravity * Time.fixedDeltaTime * downHit.normal;
+            }
```

Comment:
```suggestion
            if (isOnSlope) 
            {
                _rigidbody.AddForce(_maxSpeed * _acceleration * Time.fixedDeltaTime * directionOnSlope);
                
                // Apply gravity but perpendicular to the slope, to prevent sliding
                _rigidbody.velocity += (_actualGravityScale - 1) * _gravity * Time.fixedDeltaTime * downHit.normal;
            }
```